File: BE/src/model/listeningContent.model.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IListeningContent extends Document {
  title: string;
  description: string;
  topic: string;
  level: string;
  audioSource: 'upload' | 'tts';
  audioUrl: string;
  transcript?: string;
  duration?: number;
  thumbnailUrl?: string;
  exercises: mongoose.Types.ObjectId[];
  createdBy: mongoose.Types.ObjectId;
  isPublished: boolean;
  playCount: number;
}

const listeningContentSchema: Schema = new Schema(
  {
    title: { type: String, required: true },
    description: { type: String, default: '' },
    topic: { 
      type: String, 
      enum: ['daily_life', 'travel', 'business', 'culture', 'school', 'shopping', 'weather', 'health', 'food', 'news', 'other'],
      required: true 
    },
    level: { type: String, enum: ['N5', 'N4', 'N3', 'N2', 'N1'], required: true },
    audioSource: { type: String, enum: ['upload', 'tts'], required: true },
    // Empty until file is uploaded to Cloudinary (audioSource === 'upload')
    audioUrl: { type: String, default: '' },
    transcript: { type: String },
    duration: { type: Number },
    thumbnailUrl: { type: String },
    exercises: [{ type: Schema.Types.ObjectId, ref: 'ListeningExercise' }],
    createdBy: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    isPublished: { type: Boolean, default: false },
    playCount: { type: Number, default: 0 },
  },
  { timestamps: true }
);

export default mongoose.model<IListeningContent>('ListeningContent', listeningContentSchema);

File: mirai_mobile/lib/models/listening_content_model.dart
class ListeningContentModel {
  final String id;
  final String title;
  final String description;
  final String topic;
  final String level;
  final String audioSource;
  final String audioUrl;
  final String? transcript;
  final int? duration;
  final String? thumbnailUrl;
  final int playCount;
  final bool isPublished;
  final DateTime createdAt;

  ListeningContentModel({
    required this.id,
    required this.title,
    required this.description,
    required this.topic,
    required this.level,
    required this.audioSource,
    required this.audioUrl,
    this.transcript,
    this.duration,
    this.thumbnailUrl,
    required this.playCount,
    required this.isPublished,
    required this.createdAt,
  });

  factory ListeningContentModel.fromJson(Map<String, dynamic> json) {
    return ListeningContentModel(
      id: json['_id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      topic: json['topic'] ?? '',
      level: json['level'] ?? '',
      audioSource: json['audioSource'] ?? 'upload',
      audioUrl: json['audioUrl'] ?? '',
      transcript: json['transcript'],
      duration: json['duration'],
      thumbnailUrl: json['thumbnailUrl'],
      playCount: json['playCount'] ?? 0,
      isPublished: json['isPublished'] ?? true,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
    );
  }
}
